Add Form recognizer migration guide (#24472)
Azure/azure-sdk-for-java: sdk/formrecognizer/azure-ai-formrecognizer/migration-guide.md

> # Guide for migrating to `azure-ai-formrecognizer (4.0.0-beta.1)` from `azure-ai-formrecognizer (3.0.x)`

Reviewer: should version be 3.X.X ? or at least 3.1.X?

> This guide is intended to assist in the migration to `azure-ai-formrecognizer (4.0.0-beta.1)` from `azure-ai-formrecognizer (3.0.x)`. It will focus on side-by-side comparisons for similar operations between the two packages.

Reviewer: ```suggestion
This guide is intended to assist in the migration to `azure-ai-formrecognizer (4.0.0-beta.1)` from `azure-ai-formrecognizer (3.0.x)`. It will focus on side-by-side comparisons for similar operations between the two package versions.
```

> We assume that you are familiar with the old SDK `azure-ai-formrecognizer (3.0.x)`. If not, please refer to the new SDK README for [azure-ai-formrecognizer][README] directly rather than this migration guide.

Reviewer: I wouldn't call it old. Maybe previous?

## Table of contents

- [Guide for migrating to `azure-ai-formrecognizer (4.0.0-beta.1)` from `azure-ai-formrecognizer (3.0.x)`](#guide-for-migrating-to-azure-ai-formrecognizer-400-beta1-from-azure-ai-formrecognizer-30x)

>     - [Table of contents](#table-of-contents)

Reviewer: do we need to have these 2 ?

    - [Migration benefits](#migration-benefits)
        - [Cross Service SDK improvements](#cross-service-sdk-improvements)
    - [Important changes](#important-changes)
        - [Instantiating clients](#instantiating-clients)

>         - [Analyze Documents using a Prebuilt Model](#analyze-documents-using-a-prebuilt-model)

Reviewer: what about analyze a document vs recognizeContent?

    - [Additional samples](#additional-samples)

## Migration benefits

A natural question to ask when considering whether to adopt a new version or library is its benefits. As Azure has matured and been embraced by a more diverse group of developers, we have been focused on learning the patterns and practices to best support developer productivity and to understand the gaps that the Java
client libraries have.

There were several areas of consistent feedback expressed across the Azure client library ecosystem. The most important is that many developers have felt that the learning curve was difficult, and the APIs did not offer a good, approachable, and consistent onboarding story for those learning Azure or exploring a specific Azure service.

To improve the development experience across Azure services, a set of uniform [design guidelines][Guidelines] was created for all languages to drive a
consistent experience with established API patterns for all services. A set of [Java design guidelines][GuidelinesJava] was introduced to ensure that Java clients have a natural and idiomatic feel with respect to the Java ecosystem. Further details are available in the guidelines

> for those interested.

Reviewer 2: might just be me but this paragraph feels really general / seems like it's explaining the track 1 to track 2 migration. We changed our design to adopt the new service improvements, but I guess you could extrapolate to say that some of it was due to developer feedback.

> 

Reviewer: All of this is the same independent of 4.0.0 so I would suggest removing

> Aside from the benefits of the new design mentioned above, the `azure-ai-formrecognizer (4.0.0-beta.1)` replaces the multiple API design and adds a unified method to analyze documents with the `beginAnalyzeDocument` method and managing of operations in your form recognizer account.This version of the client library defaults to the 2021-09-30-preview version of the service. Refer to the [README][README] for more information about analyzing documents.

Reviewer 2: I would separate the thoughts about unified method and managing of operations into separate sentences. 

Since Java removed the old clients, maybe important to call that out right away too?

### Cross Service SDK improvements

> The modern Form Recognizer client library also provides the ability to share in some cross-service improvements made to the Azure development experience, such as

Reviewer 2: what is cross-service, isn't this all Form Recognizer?

Author: This was a template I have seen used in Java package, but I can try changing the section name or collapsing with the above section.

> - A unified method for extracting text and structured data from your documents across each of the client libraries.

Reviewer 2: method names? maybe mention how model_id is passed to pick the model for analysis? up to you how much detail you want to put here though

> - A unified method for analyzing analyzing text and structured data from your documents across each of the client libraries.

Reviewer 3: ```suggestion
- A unified method for analyzing text and structured data from your documents across each of the client libraries.
```

Also think about using "extract" here instead of "analyze".

> - A unified return type `DocumentModel` indicating the document type the model can analyze and the specific fields it can analyze along with the estimated confidence for each field.

Reviewer 2: nit: document types

- Specifying a modelId instead of the generated GUID when creating custom models, along with an optional description.

> - Modified `Generate Copy Authorization operation` response to return the target resource information so that it could be used directly when copying custom models method.

Reviewer 2: maybe say something about that info not needing to be supplied by user anymore?

- List Models operation now returns a paged list of prebuilt in addition to custom models and does not return in-progress and failed models.

> - Added methods for getting/listing operations useful to track the status of model creation/copying operations and any resulting errors.

Reviewer 2: mention that ops only persist for 24 hours?

> - Added methods for getting/listing operations useful to track the status of model creation/copying operations and any resulting errors.

Reviewer: I feel like these shouldn't be in the Cross Service SDK improvements, and be more like bullet points  inside the Migration benefits

## Important changes

#### Instantiating clients

In 3.x, the `FormRecognizerClient`, instantiated via the `FormRecognizerClientBuilder`. The client contains both sync and async methods.

> In 4.x, the `FormRecognizerClient` has been replaced by the `DocumentAnalysisClient`and the instantiation of the client is done through the [DocumentAnalysisClientBuilder][DocumentAnalysisClientBuilder]. The sync and async operations are separated to [DocumentAnalysisClient][DocumentAnalysisClient] and [DocumentAnalysisAsyncClient][DocumentAnalysisAsyncClient].

Reviewer 2: ```suggestion
In 4.x, the `FormRecognizerClient` has been replaced by the `DocumentAnalysisClient` and the instantiation of the client is done through the [DocumentAnalysisClientBuilder][DocumentAnalysisClientBuilder]. The sync and async operations are separated to [DocumentAnalysisClient][DocumentAnalysisClient] and [DocumentAnalysisAsyncClient][DocumentAnalysisAsyncClient].
```

<!-- embedme ./src/samples/java/com/azure/ai/formrecognizer/ReadmeSamples.java#L52-L55 -->
```java

> DocumentAnalysisClient documentAnalysisClient = new DocumentAnalysisClientBuilder()

Reviewer 2: other migration guides I've seen show the old way of doing things and then have a snippet showing the new way. I know it adds a lot to the guide, but it's a nice way to compare.

    .credential(new AzureKeyCredential("{key}"))
    .endpoint("{endpoint}")
    .buildClient();
```

> Similarly, with 4.X, the `FormTrainingClient` has been replaced by the `DocumentModelAdministrationClient`and the instantiation of the client is done through the [DocumentModelAdministrationClientBuilder][DocumentModelAdministrationClientBuilder][DocumentModelAdministrationClientBuilder]. The sync and async operations are separated to [DocumentModelAdministrationClient][DocumentModelAdministrationClient] and [DocumentModelAdministrationAsyncClient][DocumentModelAdministrationAsyncClient].

Reviewer 2: ```suggestion
Similarly, with 4.X, the `FormTrainingClient` has been replaced by the `DocumentModelAdministrationClient`and the instantiation of the client is done through the [DocumentModelAdministrationClientBuilder][DocumentModelAdministrationClientBuilder]. The sync and async operations are separated to [DocumentModelAdministrationClient][DocumentModelAdministrationClient] and [DocumentModelAdministrationAsyncClient][DocumentModelAdministrationAsyncClient].
```

<!-- embedme ./src/samples/java/com/azure/ai/formrecognizer/ReadmeSamples.java#L62-L65 -->
```java
DocumentModelAdministrationClient documentModelAdminClient = new DocumentModelAdministrationClientBuilder()
    .credential(new AzureKeyCredential("{key}"))
    .endpoint("{endpoint}")
    .buildClient();
```

#### Analyze Documents using a Prebuilt Model
Analyze data from certain types of common documents (such as receipts, invoices, business cards, or identity documents) using prebuilt models.
In 3.x, `beginRecognizeReceipts` or `beginRecognizeReceiptsFromUrl` method was used to analyze receipts.

> In 4.X, `beginRecognizeReceipts` and `beginRecognizeReceiptsFromUrl` has been replaced with `beginAnalyzeDocument` and `beginAnalyzeDocumentFromUrl` respectively. Below is an example to analyze receipt data using 4.X `beginAnalyzeDocumentFromUrl`:

Reviewer 2: maybe describe how this applies to invoices, business cards, and id docs too?

Reviewer 2: And how content is now layout? And how `language` -> `locale`?

> In 4.X, `beginRecognizeReceipts` and `beginRecognizeReceiptsFromUrl` has been replaced with `beginAnalyzeDocument` and `beginAnalyzeDocumentFromUrl` respectively. Below is an example to analyze receipt data using 4.X `beginAnalyzeDocumentFromUrl`:

Reviewer: consider adding that the return type of the method also changed

                .map(formField -> formField.getValueMap())
                .forEach(formFieldMap -> formFieldMap.forEach((key, formField) -> {
                    if ("Name".equals(key)) {
                        if (DocumentFieldType.STRING == formField.getType()) {
                            String name = formField.getValueString();
                            System.out.printf("Name: %s, confidence: %.2fs%n",
                                name, formField.getConfidence());
                        }
                    }
                    if ("Quantity".equals(key)) {
                        if (DocumentFieldType.FLOAT == formField.getType()) {
                            Float quantity = formField.getValueFloat();
                            System.out.printf("Quantity: %f, confidence: %.2f%n",
                                quantity, formField.getConfidence());
                        }
                    }
                }));
        }
    }
}
```

#### Build a custom document analysis model
In 3.X, creating a custom model required specifying `useTrainingLabels` to indicate if the model creation was using the custom labeling or not with the `beginTraining` method.

> In 4.X, we introduced the new general document model (prebuilt-document) consisting of key-value pair extraction, it is no longer necessary to train a custom model to extract key-value pairs. Thus, we removed the option to train custom models without labels. Hence, with 4.X, we can build a custom document analysis model using the `beginBuildModel` method as shown in the example below:

Reviewer 2: we never exposed key-value pairs in v3 - I would instead say something about the "train without labels" scenario

> In 4.X, we introduced the new general document model (prebuilt-document) consisting of key-value pair extraction, it is no longer necessary to train a custom model to extract key-value pairs. Thus, we removed the option to train custom models without labels. Hence, with 4.X, we can build a custom document analysis model using the `beginBuildModel` method as shown in the example below:

Reviewer: ```suggestion
In 4.X, we introduced the new prebuilt document model (prebuilt-document) consisting of key-value pair extraction, it is no longer necessary to train a custom model to extract key-value pairs. Thus, we removed the option to train custom models without labels. 
With 4.X, we can build a custom document analysis model using the `beginBuildModel` method as shown in the example below:
```

<!-- embedme ./src/samples/java/com/azure/ai/formrecognizer/ReadmeSamples.java#L293-L315 -->
```java
// Build custom document analysis model
String trainingFilesUrl = "{SAS_URL_of_your_container_in_blob_storage}";
// The shared access signature (SAS) Url of your Azure Blob Storage container with your forms.

> SyncPoller<DocumentOperationResult, DocumentModel> buildOperationPoller =

Reviewer: @kristapratico  added great comments in the Python PR https://github.com/Azure/azure-sdk-for-python/pull/20960/files that apply here as well

    documentModelAdminClient.beginBuildModel(trainingFilesUrl,
        "my-build-model",
        new BuildModelOptions().setDescription("model desc"),
        Context.NONE);

DocumentModel documentModel = buildOperationPoller.getFinalResult();

// Model Info
System.out.printf("Model ID: %s%n", documentModel.getModelId());
System.out.printf("Model Description: %s%n", documentModel.getDescription());
System.out.printf("Model created on: %s%n%n", documentModel.getCreatedOn());
documentModel.getDocTypes().forEach((key, docTypeInfo) -> {
    System.out.printf("Document type: %s%n", key);
    docTypeInfo.getFieldSchema().forEach((name, documentFieldSchema) -> {
        System.out.printf("Document field: %s%n", name);
        System.out.printf("Document field type: %s%n", documentFieldSchema.getType().toString());
        System.out.printf("Document field confidence: %.2f%n", docTypeInfo.getFieldConfidence().get(name));
    });
});
```

### Manage models

> In 3.X, listing models returned only the custom trained models using the `listCustomModel` method.

Reviewer: we should also call out how these methods now only return models that are ready to be consumed, while before models in whatever state where returned

> With 4.X, list Models operation `listModels` now returns a paged list of prebuilt in addition to custom models and does not return in-progress and failed models.

Reviewer 2: since we don't have the concept of model status anymore, maybe say something like models built successfully?

System.out.println("We have following models in the account:");
customDocumentModels.forEach(documentModelInfo -> {
    System.out.printf("Model ID: %s%n", documentModelInfo.getModelId());
    modelId.set(documentModelInfo.getModelId());

    // get custom document analysis model info
    DocumentModel documentModel = documentModelAdminClient.getModel(documentModelInfo.getModelId());
    System.out.printf("Model ID: %s%n", documentModel.getModelId());
    System.out.printf("Model Description: %s%n", documentModel.getDescription());
    System.out.printf("Model created on: %s%n", documentModel.getCreatedOn());
    documentModel.getDocTypes().forEach((key, docTypeInfo) -> {
        docTypeInfo.getFieldSchema().forEach((field, documentFieldSchema) -> {
            System.out.printf("Field: %s", field);
            System.out.printf("Field type: %s", documentFieldSchema.getType());
            System.out.printf("Field confidence: %.2f", docTypeInfo.getFieldConfidence().get(field));
        });
    });
});

// Delete Model
documentModelAdminClient.deleteModel(modelId.get());
```

#### Analyze Documents using a Custom model

> Analyze data from custom documents using your own custom models, built with your own data, so they're tailored to your documents.

Reviewer: consider adding how return types also changed

> In 3.x, `beginRecognizeCustomForms` or `beginRecognizeCustomFormsFromUrl` method was used to analyze receipts.

Reviewer 2: ```suggestion
In 3.x, `beginRecognizeCustomForms` or `beginRecognizeCustomFormsFromUrl` method was used to analyze custom documents.
```

> In 3.x, `beginRecognizeCustomForms` or `beginRecognizeCustomFormsFromUrl` method was used to analyze receipts.

Reviewer: ```suggestion
In 3.x, `beginRecognizeCustomForms` or `beginRecognizeCustomFormsFromUrl` method was used to recognize forms.
```

> In 4.X, `beginRecognizeCustomForms` and `beginRecognizeCustomFormsFromUrl` has been replaced with `beginAnalyzeDocument` and `beginAnalyzeDocumentFromUrl` respectively. Below is an example to analyze custom document data using 4.X `beginAnalyzeDocumentFromUrl`:

Reviewer: ```suggestion
In 4.X, `beginRecognizeCustomForms` and `beginRecognizeCustomFormsFromUrl` has been replaced with `beginAnalyzeDocument` and `beginAnalyzeDocumentFromUrl` respectively. Below is an example to analyze a document with a custom model using 4.X `beginAnalyzeDocumentFromUrl`:
```

> <!-- embedme ./src/samples/java/com/azure/ai/formrecognizer/ReadmeSamples.java#L83-L102 -->

Reviewer 2: I think it's worth describing the different return types - `AnalyzeResult` vs List of `RecognizedForm` and maybe different options here like `pages` now a string, no more `include_field_elements`, etc

# Guide for migrating to `azure-ai-formrecognizer (4.0.0-beta.1 - above)` from `azure-ai-formrecognizer (3.1.x - below)`

This guide is intended to assist in the migration to `azure-ai-formrecognizer (4.0.0-beta.1 - above)` from `azure-ai-formrecognizer (3.1.x - below)`. It will focus on side-by-side comparisons for similar operations between the two package versions.

> We assume that you are familiar with the previous SDK `azure-ai-formrecognizer (3.1.x - below)`. If not, please refer to the new SDK README for [azure-ai-formrecognizer][README] directly rather than this migration guide.

Reviewer 4: Maybe this would be better like this?
```suggestion
We assume that you are familiar with the previous SDK `azure-ai-formrecognizer (3.1.x - below)`. If you are new to this library, please refer to the SDK README for [azure-ai-formrecognizer][README] directly rather than this migration guide.
```

## Table of contents
- [Migration benefits](#migration-benefits)
- [Important changes](#important-changes)
    - [Instantiating clients](#instantiating-clients)
    - [Analyze documents](#analyze-documents)
      - [Using a prebuilt Model](#using-a-prebuilt-model)
      - [Using a layout model](#using-a-layout-model)
      - [Using custom model](#using-a-custom-model)
    - [Manage models](#manage-models)
- [Additional samples](#additional-samples)

## Migration benefits

A natural question to ask when considering whether to adopt a new version of the library is what the benefits of
doing so would be. As Azure Form Recognizer has matured and been embraced by a more diverse group of developers, 
we have been focused on learning the patterns and practices to best support developer productivity and add value to our 
customers.

> To improve the development experience and addressing the consistent feedback across the Form Recognizer SDK, this new 

Reviewer 4: Nit
```suggestion
To improve the development experience and address the consistent feedback across the Form Recognizer SDK, this new 
```

version of the library replaces the previously existing clients `FormRecognizerClient` and `FormTrainingClient` with
`DocumentAnalysisClient` and the `DocumentModelAdministrationClient` that provide unified methods for 

> analyzing documents and providing support for the new features added by the service in 

Reviewer 4: Small nit
```suggestion
analyzing documents and provide support for the new features added by the service in 
```

|-|-
|2021-09-30-preview | DocumentAnalysisClient and DocumentModelAdministrationClient
|2.1 | FormRecognizerClient and FormTrainingClient
|2.0 | FormRecognizerClient and FormTrainingClient

The newer Form Recognizer client library also provides the ability to share in some improvements made to the Azure development experience, such as:

- A unified method, `beginAnalyzeDocument` and `beginAnalyzeDocumentFromUrl`, for analyzing text and structured data from documents.
This method uses a `modelId` parameter for specifying the type of analysis to perform. 
The newly introduced method return type `AnalyzeResult` removes hierarchical dependencies between the previously known `FormElements`
and move them to a more top level and easily accessible position such as `AnalyzeResult.tables` instead of `RecognizedForm.pages.tables`.
The service has further matured to define cross-page elements by using the `BoundingRegion` model and by specifying the content and span information on document fields.
- A unified return type `DocumentModel` indicating the document types the model can analyze and the specific fields it can analyze along with the estimated confidence for each field.

> - Specifying a modelId instead of the generated GUID when creating custom models, along with an optional description. See [here][service_supported_models], for the supported model types.

Reviewer 4: ```suggestion
- Specifying a modelId instead of the generated GUID when creating, composing or copying models, along with an optional description. See [here][service_supported_models], for the supported model types.
```

        System.out.printf("Cell '%s', has row index %d and column index %d.%n",
            documentTableCell.getContent(),
            documentTableCell.getRowIndex(), documentTableCell.getColumnIndex());
    });
    System.out.println();
}

// Entities
analyzeResult.getEntities().forEach(documentEntity -> {
    System.out.printf("Entity category : %s, sub-category %s%n: ",
        documentEntity.getCategory(), documentEntity.getSubCategory());
    System.out.printf("Entity content: %s%n: ", documentEntity.getContent());
    System.out.printf("Entity confidence: %.2f%n", documentEntity.getConfidence());
});

// Key-value
analyzeResult.getKeyValuePairs().forEach(documentKeyValuePair -> {
    System.out.printf("Key content: %s%n", documentKeyValuePair.getKey().getContent());
    System.out.printf("Key content bounding region: %s%n",
        documentKeyValuePair.getKey().getBoundingRegions().toString());

    System.out.printf("Value content: %s%n", documentKeyValuePair.getValue().getContent());
    System.out.printf("Value content bounding region: %s%n", documentKeyValuePair.getValue().getBoundingRegions().toString());
});
```

#### Build a custom document analysis model

> - In 3.x.x, creating a custom model required specifying `useTrainingLabels` to indicate whether to use labeled data when creating the custom model with the `beginTraining` method.

Reviewer 4: Should this sentence also specifically call out that in 4.x.x it's always labeled data?

We assume that you are familiar with the previous SDK `azure-ai-formrecognizer (3.1.x - below)`. If not, please refer to the new SDK README for [azure-ai-formrecognizer][README] directly rather than this migration guide.

## Table of contents
- [Migration benefits](#migration-benefits)
- [Important changes](#important-changes)
    - [Instantiating clients](#instantiating-clients)
    - [Analyze documents](#analyze-documents)
      - [Using a prebuilt Model](#using-a-prebuilt-model)
      - [Using a layout model](#using-a-layout-model)
      - [Using custom model](#using-a-custom-model)
    - [Manage models](#manage-models)
- [Additional samples](#additional-samples)

## Migration benefits

A natural question to ask when considering whether to adopt a new version of the library is what the benefits of
doing so would be. As Azure Form Recognizer has matured and been embraced by a more diverse group of developers, 
we have been focused on learning the patterns and practices to best support developer productivity and add value to our 
customers.

To improve the development experience and addressing the consistent feedback across the Form Recognizer SDK, this new 

> version of the library replaces the previously existing clients `FomRecognizerClient` and `FormTrainingClient` with

Reviewer 2: ```suggestion
version of the library replaces the previously existing clients `FormRecognizerClient` and `FormTrainingClient` with
```

and `DocumentModelAdministrationAsyncClient`, instantiated via the [DocumentModelAdministrationClientBuilder][DocumentModelAdministrationClientBuilder].
The sync and async operations are separated to [DocumentModelAdministrationClient][DocumentModelAdministrationClient] and [DocumentModelAdministrationAsyncClient][DocumentModelAdministrationAsyncClient].

Instantiating FormRecognizerClient client with 3.x.x:
```java
FormTrainingClient formTrainingClient = new FormTrainingClientBuilder()
    .credential(new AzureKeyCredential("{key}"))
    .endpoint("{endpoint}")
    .buildClient();
```

Instantiating DocumentModelAdministrationClient client with 4.x.x:
<!-- embedme ./src/samples/java/com/azure/ai/formrecognizer/ReadmeSamples.java#L62-L65 -->
```java
DocumentModelAdministrationClient documentModelAdminClient = new DocumentModelAdministrationClientBuilder()
    .credential(new AzureKeyCredential("{key}"))
    .endpoint("{endpoint}")
    .buildClient();
```

#### Analyze documents
Analyze data from certain types of common documents (such as receipts, invoices, business cards, or identity documents) 

> using prebuilt models or from custom documents using your own custom models and extract table and structured layout data.

Reviewer 2: this sentence seems out of place. Maybe just say something about the differences between versions here?

With 4.x.x, the unified method, `beginAnalyzeDocument` and `beginAnalyzeDocumentFromUrl`:
- accepts a string type `modelId` to be any of the prebuilt model IDs or a custom model ID.

> - the return type `AnalyzeResult` model now exposes document elements, such as key-value pairs, entities, tables,

Reviewer 2: to keep with the tense of the first sentence

```suggestion
- returns the `AnalyzeResult` model which now exposes document elements, such as key-value pairs, entities, tables,
```

  document fields and values at the top level of the returned model. As compared to the previously returned model
  `RecognizedForm` which included hierarchical relationships between `FormElements` for instance tables were an element
  of a `FormPage` and not a top-level element.
- provides the functionality of `beginRecognizeCustomForms`, `beginRecognizeContent`, `beginRecognizeReceipt`,
  `beginRecognizeReceipts`, `beginRecognizeInvoices` `beginRecognizeIdentityDocuments` and `beginRecognizeBusinessCards` from the previous (azure-ai-formrecognizer 3.1.X - below) package versions.
- accepts unified `AnalyzeDocumentOptions` to specify pages and locale information for the outgoing request
- the `includeFieldElements` parameter is not supported with the `DocumentAnalysisClient`, text details are automatically included with API version `2021-09-30-preview` and later.
- the `readingOrder` parameter does not exist as the service uses `natural` reading order for the returned data.

#### Using a prebuilt model
- In 3.x.x, `beginRecognizeReceipts` and `beginRecognizeReceiptsFromUrl` method was used to analyze receipts.

> - In 4.x.x, `beginRecognizeReceipts` and `beginRecognizeReceiptsFromUrl` has been replaced with `beginAnalyzeDocument` and `beginAnalyzeDocumentFromUrl` respectively.

Reviewer 2: mention that this applies to all prebuilts, not just receipts?

Reviewer 4: Yeah I agree it's better to make that clearer

                .map(documentField -> documentField.getValueMap())
                .forEach(documentFieldMap -> documentFieldMap.forEach((key, documentField) -> {
                    if ("Name".equals(key)) {
                        if (DocumentFieldType.STRING == documentField.getType()) {
                            String name = documentField.getValueString();
                            System.out.printf("Name: %s, confidence: %.2fs%n",
                                name, documentField.getConfidence());
                        }
                    }
                    if ("Quantity".equals(key)) {
                        if (DocumentFieldType.FLOAT == documentField.getType()) {
                            Float quantity = documentField.getValueFloat();
                            System.out.printf("Quantity: %f, confidence: %.2f%n",
                                quantity, documentField.getConfidence());
                        }
                    }
                }));
        }
    }
}
```
#### Using a layout model
- Recognize text, table structures and selection marks like radio buttons and check boxes, along with their bounding box

> coordinates, from documents.

Reviewer 2: I would just say the differences here again. I think in Python we say something like "Differences between versions". I don't think we need to introduce stuff like in the readme.
